File: backend/app/schemas/integrations.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.enums import IndustryCode


class ShopifyConnectRequest(BaseModel):
    shop_domain: str = Field(..., description="e.g. my-store or my-store.myshopify.com")
    access_token: str = Field(..., min_length=10, description="Admin API access token")
    target_industry: IndustryCode = Field(
        ..., description="Industry that synced products/SKUs/sales land in"
    )
    # Optional: the custom app's API secret key. Needed only to verify inbound
    # webhooks (real-time sales). Stored encrypted; safe to omit for polling.
    api_secret: str | None = Field(default=None, min_length=8)
    sync_products: bool = True
    sync_inventory: bool = True
    sync_orders: bool = True


class SyncScope(BaseModel):
    sync_products: bool = True
    sync_inventory: bool = True
    sync_orders: bool = True


class IntegrationStatus(BaseModel):
    provider: str
    connected: bool
    status: str
    shop_domain: str | None = None
    target_industry: str | None = None
    last_sync_at: datetime | None = None
    last_sync_status: str | None = None
    last_sync_stats: dict = Field(default_factory=dict)
    error_message: str | None = None
    # Echoed shop name from the most recent validation/sync, if known.
    shop_name: str | None = None


class SyncAccepted(BaseModel):
    status: str
    detail: str


class LiveSaleRow(BaseModel):
    sale_time: datetime
    sku_code: str | None = None
    description: str | None = None
    units: int
    revenue: float | None = None
    order_id: str | None = None


class LiveSalesSummary(BaseModel):
    connected: bool
    today_units: int = 0
    today_revenue: float = 0.0
    today_orders: int = 0
    last_sale_at: datetime | None = None
    # 24 hourly unit buckets, oldest → newest.
    sparkline_hourly: list[int] = Field(default_factory=list)
    recent: list[LiveSaleRow] = Field(default_factory=list)


# ── Connector Hub (catalog of all providers) ─────────────────────────────────
class AuthFieldOut(BaseModel):
    key: str
    label: str
    type: str
    required: bool
    placeholder: str | None = None
    help: str | None = None
    options: list[str] | None = None
    secret: bool = False


class ConnectorOut(BaseModel):
    """A catalog provider plus this tenant's connection state for it."""

    key: str
    name: str
    category: str
    availability: str  # live | beta | planned
    summary: str
    feeds: list[str] = Field(default_factory=list)
    auth_fields: list[AuthFieldOut] = Field(default_factory=list)
    docs_url: str | None = None
    icon: str
    accent: str
    # Per-tenant connection state (None when never connected).
    status: str = "disconnected"  # connected | syncing | error | requested | disconnected
    connected: bool = False
    last_sync_at: datetime | None = None
    last_sync_status: str | None = None
    error_message: str | None = None
    # Push token, returned ONCE right after connecting a push provider
    # (rest_api / webhooks). Never persisted in plaintext or echoed again.
    push_token: str | None = None
    # Whether POST /integrations/{key}/sync is available for this provider
    # (direct-SQL connectors today; others sync automatically or push in).
    supports_sync: bool = False


class CategoryGroupOut(BaseModel):
    category: str
    label: str
    connectors: list[ConnectorOut]


class CatalogOut(BaseModel):
    groups: list[CategoryGroupOut]
    # Flat counts for the Hub header.
    total: int = 0
    live: int = 0
    connected: int = 0


class GenericConnectRequest(BaseModel):
    """Connect (or request) any catalog provider. ``credentials`` keys must match
    the provider's ``auth_fields``; secret values are encrypted at rest."""

    target_industry: IndustryCode = Field(
        ..., description="Industry that data from this source lands in"
    )
    credentials: dict[str, str] = Field(default_factory=dict)


class UploadResult(BaseModel):
    provider: str
    kind: str  # sales | inventory | products
    rows_total: int
    rows_imported: int
    rows_skipped: int
    products_created: int = 0
    skus_created: int = 0
    inventory_rows: int = 0
    sales_rows: int = 0
    errors: list[str] = Field(default_factory=list)

File: frontend/src/api/integrations.ts
import { apiClient } from "@/api/client";
import type { IndustryCode } from "@/types/api";

export interface IntegrationStatus {
  provider: string;
  connected: boolean;
  status: "connected" | "syncing" | "error" | "disconnected" | string;
  shop_domain?: string | null;
  target_industry?: string | null;
  last_sync_at?: string | null;
  last_sync_status?: string | null;
  last_sync_stats: Record<string, number | string | null>;
  error_message?: string | null;
  shop_name?: string | null;
}

export interface ShopifyConnectRequest {
  shop_domain: string;
  access_token: string;
  target_industry: IndustryCode;
  sync_products: boolean;
  sync_inventory: boolean;
  sync_orders: boolean;
}

export interface SyncScope {
  sync_products: boolean;
  sync_inventory: boolean;
  sync_orders: boolean;
}

export interface LiveSaleRow {
  sale_time: string;
  sku_code?: string | null;
  description?: string | null;
  units: number;
  revenue?: number | null;
  order_id?: string | null;
}

export interface LiveSalesSummary {
  connected: boolean;
  today_units: number;
  today_revenue: number;
  today_orders: number;
  last_sale_at?: string | null;
  sparkline_hourly: number[];
  recent: LiveSaleRow[];
}

// ── Connector Hub (catalog of all providers) ────────────────────────────────
export interface AuthField {
  key: string;
  label: string;
  type: string;
  required: boolean;
  placeholder?: string | null;
  help?: string | null;
  options?: string[] | null;
  secret: boolean;
}

export interface Connector {
  key: string;
  name: string;
  category: string;
  availability: "live" | "beta" | "planned";
  summary: string;
  feeds: string[];
  auth_fields: AuthField[];
  docs_url?: string | null;
  icon: string;
  accent: string;
  status: "connected" | "syncing" | "error" | "requested" | "disconnected" | string;
  connected: boolean;
  last_sync_at?: string | null;
  last_sync_status?: string | null;
  error_message?: string | null;
  // Returned once, right after connecting a push provider (rest_api / webhooks).
  push_token?: string | null;
  supports_sync?: boolean;
}

export interface CategoryGroup {
  category: string;
  label: string;
  connectors: Connector[];
}

export interface Catalog {
  groups: CategoryGroup[];
  total: number;
  live: number;
  connected: number;
}

export interface GenericConnectRequest {
  target_industry: IndustryCode;
  credentials: Record<string, string>;
}

export interface UploadResult {
  provider: string;
  kind: string;
  rows_total: number;
  rows_imported: number;
  rows_skipped: number;
  products_created: number;
  skus_created: number;
  inventory_rows: number;
  sales_rows: number;
  errors: string[];
}

export const integrationsApi = {
  // Hub catalog
  catalog: () => apiClient.get<Catalog>("/integrations/catalog").then((r) => r.data),
  connect: (provider: string, body: GenericConnectRequest) =>
    apiClient.post<Connector>(`/integrations/${provider}/connect`, body).then((r) => r.data),
  disconnect: (provider: string) =>
    apiClient.delete<Connector>(`/integrations/${provider}`).then((r) => r.data),
  sync: (provider: string) =>
    apiClient.post<{ status: string; detail: string }>(`/integrations/${provider}/sync`, {}).then((r) => r.data),
  upload: (file: File, kind: string, targetIndustry: IndustryCode) => {
    const form = new FormData();
    form.append("file", file);
    form.append("kind", kind);
    form.append("target_industry", targetIndustry);
    return apiClient
      .post<UploadResult>("/integrations/upload", form, {
        headers: { "Content-Type": "multipart/form-data" },
      })
      .then((r) => r.data);
  },

  // Shopify (dedicated flow)
  shopifyStatus: () =>
    apiClient.get<IntegrationStatus>("/integrations/shopify").then((r) => r.data),
  shopifyConnect: (body: ShopifyConnectRequest) =>
    apiClient
      .post<IntegrationStatus>("/integrations/shopify/connect", body)
      .then((r) => r.data),
  shopifySync: (body: SyncScope) =>
    apiClient
      .post<{ status: string; detail: string }>("/integrations/shopify/sync", body)
      .then((r) => r.data),
  shopifyDisconnect: () =>
    apiClient.delete<IntegrationStatus>("/integrations/shopify").then((r) => r.data),
  shopifyLive: () =>
    apiClient.get<LiveSalesSummary>("/integrations/shopify/live").then((r) => r.data),
};
